projectq/meta/_control.py

```diff
@@ -24,18 +24,63 @@
 """
 
 from projectq.cengines import BasicEngine
-from projectq.meta import ComputeTag, UncomputeTag
+from projectq.meta import ComputeTag, UncomputeTag, Compute, Uncompute
 from projectq.ops import ClassicalInstructionGate
 from projectq.types import BasicQubit
+from projectq.ops import X
 from ._util import insert_engine, drop_engine_after
+from projectq.ops import CtrlAll
+
+
+def canonical_ctrl_state(ctrl_state, num_qubits):
+    """
+    Return canonical form for control state
+
+    Args:
+        ctrl_state (int,str,CtrlAll): Initial control state representation
+        num_qubits (int): number of control qubits
+
+    Returns:
+        Canonical form of control state (currently a string composed of '0' and '1')
+
+    Note:
+        In case of integer values for `ctrl_state`, the least significant bit applies to the first qubit in the qubit
+        register, e.g. if ctrl_state == 2, its binary representation if '10' with the least significan bit being 0.
+        This means in particular that the followings are equivalent:
+
+        .. code-block:: python
+
+            canonical_ctrl_state(6, 3) == canonical_ctrl_state(6, '110')
+    """
+    if not num_qubits:
+        return ''
+
+    if isinstance(ctrl_state, CtrlAll):
+        if ctrl_state == CtrlAll.One:
+            return '1' * num_qubits
+        return '0' * num_qubits
+
+    if isinstance(ctrl_state, int):
+        # If the user inputs an integer, convert it to binary bit string
+        converted_str = '{0:b}'.format(ctrl_state).zfill(num_qubits)[::-1]
+        assert len(converted_str) == num_qubits, 'Control state has different length than control qubits'
+        return converted_str
```

Comment:
Why did you re-introduce this check? This will never ever get triggered since we call the `zfill(...)` method that effectively *guarantees* that the string will always have exactly `num_qubits` characters.

Reply:
Ok, forgot about cases like `canonical_ctrl_state(8, 2)` which would give out '001' as canonical form and is wrong for only 2 qubits.